packages/mattermost-redux/src/actions/posts.ts

```diff
@@ -930,22 +953,48 @@ export function getProfilesAndStatusesForPosts(postsArrayOrMap: Post[]|Map<strin
         return Promise.resolve();
     }
 
-    const posts = Object.values(postsArrayOrMap);
+    const postsArray: Post[] = Object.values(postsArrayOrMap);
 
-    if (posts.length === 0) {
+    if (postsArray.length === 0) {
         return Promise.resolve();
     }
 
+    const postsDictionary: Dictionary<Post> = {};
+    for (let i = 0; i < postsArray.length; i++) {
+        postsDictionary[postsArray[i].id] = postsArray[i];
+    }
+
     const state = getState();
+    let posts: IDMappedObjects<Post> = {};
+    if (state.entities.posts?.posts) {
+        posts = state.entities.posts.posts;
+    }
     const {currentUserId, profiles, statuses} = state.entities.users;
 
     // Statuses and profiles of the users who made the posts
     const userIdsToLoad = new Set<string>();
     const statusesToLoad = new Set<string>();
+    const postsToLoad = new Set<string>();
 
-    Object.values(posts).forEach((post) => {
+    postsArray.forEach((post) => {
         const userId = post.user_id;
 
+        if (post.metadata && post.metadata.embeds) {
+            post.metadata.embeds.forEach((embed: any) => {
+                if (embed.type === 'permalink' && embed.data) {
+                    if (embed.data.user_id && !profiles[embed.data.user_id] && embed.data.user_id !== currentUserId) {
+                        userIdsToLoad.add(embed.data.user_id);
+                    }
+                    if (embed.data.id && posts && !posts[embed.data.id] && !postsDictionary[embed.data.id]) {
+                        postsToLoad.add(embed.data.id);
```

Comment:
Loading these posts asynchronously isn't going to work since I assume the previews aren't a fixed height.

We've had a long history fighting with issues caused by posts changing their height after they are first rendered. It causes a lot of problems around positioning the user in the right place when first loading (such as ensuring the New Messages line is on screen), keeping them in the right place later on (like at the bottom of the channel), and ensuring infinite scrolling works smoothly. Post metadata was originally added mostly to ensure the post components had all the information they needed to size themselves correctly and prevent changing sizes.

I assume this was being done this way to solve the permission issues we identified around not sending posts a user didn't have permissions too, but unfortunately, I think that needs to be addressed on the server side.

Reply:
We decided to add the whole post object to the metadata. When the webapp receives the posts array we check metadata for each post and if we find a permalink we add the embedded post to the post redux store and remove it from the post object from the metadata (metadata retains the post_id so we can find the post in our redux store).

Reply:
Awesome, that sounds perfect to me.